src/xrpld/app/tx/detail/LoanSet.cpp

```diff
@@ -234,6 +234,25 @@ LoanSet::preclaim(PreclaimContext const& ctx)
         // Should be impossible
         return tefBAD_LEDGER;  // LCOV_EXCL_LINE
     Asset const asset = vault->at(sfAsset);
+
+    if (auto const err =
+            requireAuth(ctx.view, asset, borrower, AuthType::WeakAuth))
+    {
+        JLOG(ctx.j.warn()) << "Borrower not authorized to hold asset";
+        return tecNO_AUTH;
+    }
+
+    if (auto const originationFee = tx[~sfLoanOriginationFee];
+        originationFee && *originationFee != Number())
+    {
+        if (auto const err =
+                requireAuth(ctx.view, asset, brokerOwner, AuthType::WeakAuth))
+        {
+            JLOG(ctx.j.warn()) << "Broker not authorized to receive fee";
+            return tecNO_AUTH;
+        }
+    }
```

Comment:
Should `addEmptyHolding` call `requireAuth`?

Reply:
That's ok, as long as the failure in `doApply` claims a fee.

`preclaim` was split out of `doApply` exists as sort of a shortcut to save effort, and avoid failing after doing a lot of work, but when the shortcut causes more work, or just doesn't make sense, it's perfectly fine to do the check in `doApply`. IMHO.